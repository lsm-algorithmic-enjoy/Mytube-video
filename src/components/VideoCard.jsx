import React from "react";
import { formatAgo } from "../util/date";

export default function VideoCard({ video }) {
  const { title, thumbnails, channelTitle, publishedAt } = video.snippet;
  return (
    <li>
      <img src={thumbnails.medium.url} alt={title} />
      <div>
        <p>{title}</p>
        <p>{channelTitle}</p>
        <p>{formatAgo(publishedAt, "ko")}</p>
      </div>
    </li>
  );
}
