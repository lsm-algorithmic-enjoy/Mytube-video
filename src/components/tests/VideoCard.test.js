import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import { Route, useLocation } from "react-router-dom";
import { withRouter } from "../../tests/utils";
import { fakeVideo as video } from "../../tests/videos";
import { formatAgo } from "../../util/date";
import VideoCard from "../VideoCard";

describe("VideoCard", () => {
  const { title, channelTitle, publishedAt, thumbnails } = video.snippet;

  it("renders grid type correctly", () => {
    const component = renderer.create(
      withRouter(<Route path="/" element={<VideoCard video={video} />} />)
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders list type correctly", () => {
    const component = renderer.create(
      withRouter(
        <Route path="/" element={<VideoCard video={video} type="list" />} />
      )
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders video item", () => {
    render(
      withRouter(<Route path="/" element={<VideoCard video={video} />} />)
    );
    const image = screen.getByRole("img");
    expect(image.src).toBe(thumbnails.medium.url);
    expect(image.alt).toBe(title);
    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(channelTitle)).toBeInTheDocument();
    expect(screen.getByText(formatAgo(publishedAt))).toBeInTheDocument();
  });

  it("navigates to detailed video page with video state when clicked", () => {
    function LocationStateDisplay() {
      return <pre>{JSON.stringify(useLocation().state)}</pre>;
    }
    render(
      withRouter(
        <>
          <Route path="/" element={<VideoCard video={video} />} />
          <Route
            path={`/videos/watch/${video.id}`}
            element={<LocationStateDisplay />}
          />
        </>
      )
    );
    const card = screen.getByRole("listitem");
    userEvent.click(card);

    expect(screen.getByText(JSON.stringify({ video }))).toBeInTheDocument();
  });
});
