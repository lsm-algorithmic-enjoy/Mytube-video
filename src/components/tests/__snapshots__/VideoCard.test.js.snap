// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard renders grid type correctly 1`] = `
<li
  className=""
  onClick={[Function]}
>
  <img
    alt="title"
    className="w-full"
    src="http://image/"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      title
    </p>
    <p
      className="text-sm opacity-80"
    >
      channelTitle
    </p>
    <p
      className="text-sm opacity-60"
    >
      just now
    </p>
  </div>
</li>
`;

exports[`VideoCard renders list type correctly 1`] = `
<li
  className="flex gap-1 m-2"
  onClick={[Function]}
>
  <img
    alt="title"
    className="w-60 mr-2"
    src="http://image/"
  />
  <div>
    <p
      className="font-semibold my-2 line-clamp-2"
    >
      title
    </p>
    <p
      className="text-sm opacity-80"
    >
      channelTitle
    </p>
    <p
      className="text-sm opacity-60"
    >
      just now
    </p>
  </div>
</li>
`;
